services/venice-controller/src/main/java/com/linkedin/venice/controller/StoreBackupVersionCleanupService.java

```diff
@@ -224,34 +228,29 @@ private boolean validateAllServerOnCurrentVersion(Store store, String clusterNam
    * @return whether any backup version is removed or not
    */
   protected boolean cleanupBackupVersion(Store store, String clusterName) {
-    if (!whetherStoreReadyToBeCleanup(store, defaultBackupVersionRetentionMs, time)) {
+    int currentVersionNum = store.getCurrentVersion();
+    Version currentVersion = store.getVersion(currentVersionNum).get();
+    boolean isCurrentVersionFromRepush = Version.isPushIdRePush(currentVersion.getPushJobId());
+
+    if (!isCurrentVersionFromRepush && !whetherStoreReadyToBeCleanup(store, defaultBackupVersionRetentionMs, time)) {
       // not ready to clean up backup versions yet
       return false;
     }
 
     List<Version> versions = store.getVersions();
-    List<Version> readyToBeRemovedVersions = new ArrayList<>();
-    int currentVersion = store.getCurrentVersion();
-
-    if (currentVersion == NON_EXISTING_VERSION) {
-      return false;
-    }
-
     // Do not delete version unless all routers and all servers are on same current version
     if (multiClusterConfig.getControllerConfig(clusterName).isBackupVersionMetadataFetchBasedCleanupEnabled()
-        && (!validateAllRouterOnCurrentVersion(store, clusterName, currentVersion)
-            || !validateAllServerOnCurrentVersion(store, clusterName, currentVersion))) {
+        && (!validateAllRouterOnCurrentVersion(store, clusterName, currentVersionNum)
+            || !validateAllServerOnCurrentVersion(store, clusterName, currentVersionNum))) {
```

Comment:
These two function calls are fairly expensive, so shall we make sure there are lingering backup versions before running such validation?

Reply:
Good idea, that probably skip running the service on store.versions.size == 1.